src/main/java/com/lorepo/icplayer/client/module/text/TextPrintable.java

```diff
@@ -78,97 +107,165 @@ private String makePrintableInput(String parsedText, boolean showAnswers) {
+			// it will break into lines, with no part smaller than minSplitSize
+			for (int j = 0; j < minSplitSize; j++) {
+				value += "&nbsp;";
+			}
+
+			boolean nextNbsp = false;
+			for (int j = 0; j < spaceCount - minSplitSize * 2; j++) {
+				if (nextNbsp) {
 					value += "&nbsp;";
+				} else {
+					value += " ";
 				}
+				nextNbsp = !nextNbsp;
 			}
 
-			span.setInnerHTML(value);	
-			String newValue = "&nbsp;" + span.getString();
-			
-			parsedText = parsedText.replace(oldValue, newValue);
+			for (int j = 0; j < minSplitSize; j++) {
+				value += "&nbsp;";
+			}
+		} else {
+			for (int j = 0; j < spaceCount; j++) {
+				value += "&nbsp;";
+			}
 		}
+		if (userAnswer != null && userAnswer.length() > 0 && showAnswers) {
+			do {
+				if (userAnswer.equals(answers.next())) {
+					span.addClassName("ic_text-correct-answer");
+					break;
+				}
+				if (!answers.hasNext()) {
+					span.addClassName("ic_text-wrong-answer");
+					break;
+				}
+			} while (answers.hasNext());
+		}
+
+		span.setInnerHTML(value);
+		String newValue = "&nbsp;" + span.getString();
+
+		return newValue;
+	}
 
+	private String makePrintableMathInput(String parsedText, boolean showAnswers) {
+		if (!model.hasMathGaps()) return parsedText;
+		Element replacementWrapper = DOM.createDiv();
+		String innerHTML = "";
+		replacementWrapper.addClassName("math-gap-replacement-wrapper");
+		for (GapInfo gapInfo: model.getGapInfos()) {
+			innerHTML += createPrintableGap(gapInfo, gapInfo.getPlaceHolder(), showAnswers);
+		}
+		replacementWrapper.setInnerHTML(innerHTML);
+		parsedText += replacementWrapper.getString();
 		return parsedText;
 	}
 	
 	private String makePrintableDropdowns(String parsedText, boolean showAnswers) {
 		HTML html = new HTML(parsedText);
-		
+
 		NodeList<Element> selects = html.getElement().getElementsByTagName("select");
 		for (int i = 0; i < selects.getLength(); i++) {
 			Element select = selects.getItem(i);
+			Element span = DOM.createSpan();
 			NodeList<Element> options = select.getElementsByTagName("option");
-			
+			HashMap<String, String> printableState = model.getPrintableState();
+
 			String values = "";
+			String userAnswer = "";
+			InlineChoiceInfo choiceInfo = model.getChoiceInfos().get(i);
+			if (printableState != null) {
+				userAnswer = printableState.get(choiceInfo.getId());
+			}
+
 			for (int j = 0; j < options.getLength(); j++) {
 				Element option = options.getItem(j);
 				String value = option.getInnerText();
 				if (!value.equals("---")) {
-					if (showAnswers) {
-						InlineChoiceInfo choiceInfo = model.getChoiceInfos().get(i);
-						if (choiceInfo.getAnswer().equals(value)) {
-							value = "<u>" + value + "</u>";
+					if (userAnswer == null || userAnswer.length() == 0) {
+						if (showAnswers && printableState == null) {
+							if (choiceInfo.getAnswer().equals(value)) {
+								Element underlined = DOM.createElement("u");
+								underlined.setInnerHTML(value);
+								underlined.getString();

```

Comment:
To chyba do wyrzucenia (w następnej linjce jest dopiero przypisanie do zmiennej wyniku tej samej funkcji)